### PR TITLE
Dropout

### PR DESCRIPTION
**Pull request by haticeaydinn**
Completed dropout notebook

### CODE REVIEW

File: Chapter 4. Deep Learning/Module 0. Neural Networks and PyTorch/Dropout.ipynb [hunk 1]
```diff
@@ -0,0 +1,2128 @@
+{
```
Comment (life-efficient) [inline Chapter 4. Deep Learning/Module 0. Neural Networks and PyTorch/Dropout.ipynb:1]: <p>"drop out" -&gt; remove (drop out)</p><p> </p><p>"based on probability" -&gt; "with some probability"</p><p> </p><p>We don't take the average of each model - if we did that, for each prediction we would have to take the average of potentially billions of predictions! We just don't drop out any of the neurons. But now, the model has not learnt to depend on any particular neurons, because it couldn't count on them being there during training.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core/Practical-ML-DS/pull/148/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (life-efficient) [inline Chapter 4. Deep Learning/Module 0. Neural Networks and PyTorch/Dropout.ipynb:1]: <p>How do we apply dropout in practice? For each layer in the network, we set a probability that the neuron activation will be set to zero (it will be dropped out). To implement this, we create a "mask". A mask is a binary vector which is multiplied element wise with it's corresponding layer's neuron activations. Those neurons whose corresponding mask element is zero are the ones which are dropped out, the other elements are not (they are multiplied by 1).</p><p> </p><p>Equivalently to setting the neuron activation to zero, we could just set the incoming or outgoing arc weights to zero. But in this case, we'd have to apply that to potentially many arcs, whereas when dropping out a neuron, we always just have to multiply the single neuron by zero.</p><p> </p><p>make it clear that we generate a new random mask for each input during training, and do not apply the mask during testing </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core/Practical-ML-DS/pull/148/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (life-efficient) [inline Chapter 4. Deep Learning/Module 0. Neural Networks and PyTorch/Dropout.ipynb:1]: <p>i think this section with the sports example is a bit too simple, we can probably remove it</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core/Practical-ML-DS/pull/148/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (life-efficient) [inline Chapter 4. Deep Learning/Module 0. Neural Networks and PyTorch/Dropout.ipynb:1]: <p>great gif!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core/Practical-ML-DS/pull/148/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (life-efficient) [inline Chapter 4. Deep Learning/Module 0. Neural Networks and PyTorch/Dropout.ipynb:1]: <p>not quite sure what that means.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core/Practical-ML-DS/pull/148/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (life-efficient) [inline Chapter 4. Deep Learning/Module 0. Neural Networks and PyTorch/Dropout.ipynb:1]: <p>the notation is not clear - none of those variables have been defined so the reader will not know what they mean. Can we make it any clearer?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core/Practical-ML-DS/pull/148/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (life-efficient) [inline Chapter 4. Deep Learning/Module 0. Neural Networks and PyTorch/Dropout.ipynb:1]: <p>nice</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core/Practical-ML-DS/pull/148/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (life-efficient) [inline Chapter 4. Deep Learning/Module 0. Neural Networks and PyTorch/Dropout.ipynb:1]: <p>nice</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core/Practical-ML-DS/pull/148/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

Comment (life-efficient) [inline Chapter 4. Deep Learning/Module 0. Neural Networks and PyTorch/Dropout.ipynb:1]: <p>you should only need one training function - it takes the model as an argument. Move the optimiser inside the training function.</p><p> </p><p>Also every epoch, evaluate the generalisation error by calculating the loss on the validation set</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core/Practical-ML-DS/pull/148/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='45'/>

Comment (life-efficient) [inline Chapter 4. Deep Learning/Module 0. Neural Networks and PyTorch/Dropout.ipynb:1]: <p>great!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core/Practical-ML-DS/pull/148/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='51'/>

Comment (life-efficient) [inline Chapter 4. Deep Learning/Module 0. Neural Networks and PyTorch/Dropout.ipynb:1]: <p>the first 2 sentences are confusing - this is not what we do. I'd remove it. Instead just explain more clearly that we divide the model wights by the layer;s corresponding dropout probability (not necessarily half).</p><p> </p><p>This will give -&gt; This has empirically been shown</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core/Practical-ML-DS/pull/148/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='52'/>

Comment (haticeaydinn) [reply Chapter 4. Deep Learning/Module 0. Neural Networks and PyTorch/Dropout.ipynb:1]: <p>I was watching a video about dropout (in min 8:30,  https://www.youtube.com/watch?v=D8PJAL-MZv8) and in min 8:30, it explained why we don't dropout during testing. You're right, I understood what you've meant above.</p><p> </p><p>It's theoretically possible but will be computationally inefficient! I'll edit it.</p>

File: Chapter 4. Deep Learning/Module 0. Neural Networks and PyTorch/Dropout.ipynb [hunk 2]
```diff
@@ -0,0 +1,904 @@
...
```
Comment (life-efficient) [inline Chapter 4. Deep Learning/Module 0. Neural Networks and PyTorch/Dropout.ipynb:1]: <p>where are these images?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core/Practical-ML-DS/pull/148/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (life-efficient) [inline Chapter 4. Deep Learning/Module 0. Neural Networks and PyTorch/Dropout.ipynb:1]: <p>torch actually has a dropout layer </p><p> </p><p>torch.nn.Dropout()</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core/Practical-ML-DS/pull/148/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (life-efficient) [inline Chapter 4. Deep Learning/Module 0. Neural Networks and PyTorch/Dropout.ipynb:1]: <p>the dropout model</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core/Practical-ML-DS/pull/148/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

Comment (life-efficient) [inline Chapter 4. Deep Learning/Module 0. Neural Networks and PyTorch/Dropout.ipynb:1]: <p>dont need the get_N_params</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core/Practical-ML-DS/pull/148/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='36'/>
